events/publisher.go

```diff
@@ -0,0 +1,87 @@
+package events
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/hellomd/go-sdk/rabbit"
+	"github.com/streadway/amqp"
+)
+
+const (
+	// ExchangeName is the exchange to which events are published
+	ExchangeName = "x-events"
+
+	durable    = true
+	autoDelete = false
+	internal   = false
+	noWait     = false
+	mandatory  = false
+	immediate  = false
+)
+
+// NewPublisher creates a new client that can publish events
+//
+// This publisher is concurrent safe and should be reused as much as possible
+// because of its initialization logic that involves declaring the RabbitMQ exchange.
+func NewPublisher(amqpURL string) (Publisher, error) {
+	c := &publisher{amqpURL: amqpURL}
+	if err := c.bootstrap(); err != nil {
+		return nil, fmt.Errorf("error bootstrapping events: %v", err)
+	}
+
+	return c, nil
+}
+
+type publisher struct {
+	amqpURL      string
+	rlock, wlock sync.Mutex
```

Comment:
Because `sync.RWMutex` doesn't support "promoting" a `RLock` to a `Lock`. This is the way that I know of to get an equivalent behavior.